### ISSUE TITLE
induction for truncated types

### ISSUE THREAD

**Issue by fpvandoorn** [leanprover/lean3#892]
Many induction schemes for HITs only work if the "motive" is truncated.

For example in the following recursor, `Pt` specifies that `P` must be truncated.

``` lean
trunc.rec {n : trunc_index} {A : Type} {P : trunc n A → Type}
    [Pt : Πaa, is_trunc n (P aa)] (H : Πa, P (tr a)) : Πaa, P aa
```

In the current implementation, using `induction` for such recursors doesn't work, because type class inference never fires for `Pt`, and at the end of your tactic proof, you're left with unsolved metavariables:

``` lean
open is_trunc unit

example (x : trunc -1 unit) : unit :=
begin induction x, exact star end -- fails

example (x : trunc -1 unit) : unit :=
begin refine trunc.rec _ x, intro u, exact star end -- succeeds
```

It would be nice if the `induction` tactic would fire type class resolution for arguments such as `Pt` (and failing the tactic if the type class inference fails - or, alternatively, add it as subgoal). Other examples where this is the case are [`set_quotient.rec`](https://github.com/leanprover/lean/blob/master/hott/hit/set_quotient.hlean#L29) and [`set_quotient.elim`](https://github.com/leanprover/lean/blob/master/hott/hit/set_quotient.hlean#L50). In all examples, the variable to be synthesized is the variable immediately after the motive.

**Comment by javra** [comment]
Oh yes, I run into the same situation quite often.

**Comment by leodemoura** [comment]
The induction tactic invokes the type class resolution for implicit arguments.
Could you confirm this is an issue? The following works for me

``` lean
open is_trunc unit

attribute trunc.rec [recursor 6]

example (x : trunc -1 unit) : unit :=
begin induction x, exact star end
```

**Comment by leodemoura** [comment]
BTW, in my branch, the type of `trunc.rec` is:

``` lean
trunc.rec :
  Π {n : ℕ₋₂} {A : Type} {P : trunc n A → Type} [Pt : Π (aa : trunc n A), is_trunc n (P aa)],
    (Π (a : A), P (trunc.tr a)) → (Π (aa : trunc n A), P aa)
```

**Comment by fpvandoorn** [comment]
Thanks for investigating. 

Sorry, I gave the wrong example in the OP (opening post). The example currently doesn't work because there is no `[recursor]` specified.

But I'm surprised that your example works. Similar examples didn't work "out in the wild."

I investigated a bit, and the problem seems to occur when we specify multiple recursors for a single type. Recall that for the HITs we specify both the `elim` and `rec` rules as recursors, since the arguments for the `elim` rule are slightly simpler to prove.

The following example shows the behaviour I wanted to show in the OP:

``` lean
open is_trunc unit

protected definition trunc.elim {n : trunc_index} {A : Type} {P : Type}
  [Pt : is_trunc n P] (H : A → P) : trunc n A → P :=
trunc.rec H

attribute trunc.rec  [recursor 6]
attribute trunc.elim [recursor 6]

example (x : trunc -1 unit) : unit :=
begin induction x, exact star end
```

The error I get is

```
failed to add declaration 'example' to environment, value has metavariables
remark: set 'formatter.hide_full_terms' to false to see the complete term
  λ (x : …),
    @trunc.elim … unit unit ?M_1 … x
```

Last remark: for `trunc` we could remove the `[recursor]` attribute for `trunc.elim`, since `trunc.elim` is really a special case of `trunc.rec`. But `set_quotient` has the same problem, and there the `elim` and `rec` rules are different.
